src/jit/codegenarmarch.cpp

```diff
@@ -1923,149 +1923,258 @@ void CodeGen::genCodeForCpBlkHelper(GenTreeBlk* cpBlkNode)
 {
-    // Make sure we got the arguments of the cpblk operation in the right registers
-    unsigned size    = cpBlkNode->Size();
-    GenTree* dstAddr = cpBlkNode->Addr();
-    GenTree* source  = cpBlkNode->Data();
-    GenTree* srcAddr = nullptr;
+    assert(node->OperIs(GT_STORE_BLK));
 
-    assert((size != 0) && (size <= CPBLK_UNROLL_LIMIT));
+#ifndef _TARGET_ARM64_
+    NYI_ARM("genCodeForInitBlkUnroll");
+#else
+    regNumber dstAddrBaseReg = genConsumeReg(node->Addr());
+    unsigned  dstOffset      = 0;
 
-    emitter* emit = GetEmitter();
+    regNumber srcReg;
+    GenTree*  src = node->Data();
 
-    if (dstAddr->isUsedFromReg())
+    if (src->OperIs(GT_INIT_VAL))
     {
-        genConsumeReg(dstAddr);
+        assert(src->isContained());
+        src = src->gtGetOp1();
     }
 
-    if (cpBlkNode->gtFlags & GTF_BLK_VOLATILE)
+    if (!src->isContained())
+    {
+        srcReg = genConsumeReg(src);
+    }
+    else
+    {
+        assert(src->IsIntegralConst(0));
+        srcReg = REG_ZR;
+    }
+
+    if (node->IsVolatile())
     {
-        // issue a full memory barrier before a volatile CpBlkUnroll operation
         instGen_MemoryBarrier();
     }
 
-    if (source->gtOper == GT_IND)
+    emitter* emit = GetEmitter();
+    unsigned size = node->GetLayout()->GetSize();
+
+    for (unsigned regSize = 2 * REGSIZE_BYTES; size >= regSize; size -= regSize, dstOffset += regSize)
+    {
+        emit->emitIns_R_R_R_I(INS_stp, EA_8BYTE, srcReg, srcReg, dstAddrBaseReg, dstOffset);
+    }
+
+    for (unsigned regSize = REGSIZE_BYTES; size > 0; size -= regSize, dstOffset += regSize)
     {
-        srcAddr = source->gtGetOp1();
-        if (srcAddr->isUsedFromReg())
+        while (regSize > size)
         {
-            genConsumeReg(srcAddr);
+            regSize /= 2;
+        }
+
+        instruction storeIns;
+
+        switch (regSize)
+        {
+            case 1:
+                storeIns = INS_strb;
+                break;
+            case 2:
+                storeIns = INS_strh;
+                break;
+            default:
+                storeIns = INS_str;
+                break;
+        }
+
+        emit->emitIns_R_R_I(storeIns, EA_ATTR(regSize), srcReg, dstAddrBaseReg, dstOffset);
+    }
+#endif // _TARGET_ARM64_
+}
+
+//----------------------------------------------------------------------------------
+// genCodeForCpBlkUnroll - Generate unrolled block copy code.
+//
+// Arguments:
+//    node - the GT_STORE_BLK node to generate code for
+//
+void CodeGen::genCodeForCpBlkUnroll(GenTreeBlk* node)
+{
+    assert(node->OperIs(GT_STORE_BLK));
+
+    unsigned  dstLclNum      = BAD_VAR_NUM;
+    regNumber dstAddrBaseReg = REG_NA;
+    unsigned  dstOffset      = 0;
+    GenTree*  dstAddr        = node->Addr();
+
+    if (!dstAddr->isContained())
+    {
+        dstAddrBaseReg = genConsumeReg(dstAddr);
+
+        // TODO-Cleanup: This matches the old code behavior in order to get 0 diffs. It's otherwise
```

Comment:
Yes, this will be fixed in the PR from where all this cleanup was extracted from: https://github.com/dotnet/coreclr/pull/21711/commits/10048041dbacb24726dee9cfe32111edf0456d07